cli/lib/tasks/download.js

```diff
@@ -269,9 +267,30 @@ const downloadFromUrl = ({ url, downloadDestination, progress, ca }) => {
         )
 
         reject(err)
+        // status codes here are all 2xx
+      } else {
+        // We only enable this pipe connection when we know we've got a successful return
+        // and handle the completion with verify and resolve
+        // there was a possible race condition between end of request and close of writestream
+        // that is made ordered with this Promise.all
+        Promise.all([new Promise((r) => {
+          return response.pipe(fs.createWriteStream(downloadDestination).on('close', r))
+        }), new Promise((r) => response.on('end', r))])
+        .then(() => {
+          debug('downloading finished')
+          verifyDownloadedFile(downloadDestination, expectedSize,
+            expectedChecksum)
+          .then(() => debug('verified'))
+          .then(() => resolve(version))
```

Comment:
I think we should be resolving the redirected version here.

Reply:
Then you wouldn't have to pass in version.

Reply:
actually this isn't true, you do have to pass in the version.

Reply:
recursion :)

Reply:
Actually, I think you're missing there are two different resolve calls -one here in the '200 success downloaded' point and one which is in the .then(resolve) from the downloadFromUrl in the redirect recursion point.

when there is a success we DO have the right version - Because downloadFromUrl has been called with the redirectVersion as the version field in the argument list at that point, that version gets returned from the 200 handle block which then gets resolved in the .then in the caller instance of the function.